### ISSUE TITLE
Error with `asyncio.sleep()`: Argument type is partially unknown

### ISSUE THREAD

**Issue by wch** [microsoft/pyright#3475]
I'm not sure if this is an issue with pyright or with typeshed, so please let me know if it belongs in typeshed.

With the latest version of pyright (1.1.247), this code results in an error:

```py
from typing import Awaitable, TypeVar
import asyncio
T = TypeVar("T")

async def wrap(coro: Awaitable[T]) -> T:
    return await coro

async def wrap_sleep() -> None:
    await wrap(asyncio.sleep(0))
```

The result when I run `pyright` on it:

```
Found 1 source file
/..../test/test.py
  /..../test/test.py:9:16 - error: Argument type is partially unknown
    Argument corresponds to parameter "coro" in function "wrap"
    Argument type is "Coroutine[Any, Any, Unknown]" (reportUnknownArgumentType)
1 error, 0 warnings, 0 informations 
```

However, if `result` is specified, it passes:

```py
async def wrap_sleep1() -> None:
    await wrap(asyncio.sleep(0, result=None))
```

Similarly, if I create a wrapper function for `asyncio.sleep()` where the return type is explicitly set to `None`, it passes:

```py
async def sleep0() -> None:
    await asyncio.sleep(0)

async def wrap_sleep2() -> None:
    await wrap(sleep0())
```

**Comment by JelleZijlstra** [comment]
For reference, `sleep()` is defined here: https://github.com/python/typeshed/blob/master/stdlib/asyncio/tasks.pyi#L273:

```python
async def sleep(delay: float, result: _T = ...) -> _T: ...
```

So it makes sense that pyright can't infer the return value if the `result` parameter is not provided. We should instead use overloads: one with one param that returns None, and one with two params that returns `_T`.

**Comment by erictraut** [comment]
Yeah, I agree with Jelle that the definition for `sleep` should use an overload to avoid an unsolvable TypeVar. Here's how that would look:

```python
    @overload
    async def sleep(delay: float) -> None: ...
    @overload
    async def sleep(delay: float, result: _T = ...) -> _T: ...
```

Please file a bug (or even better, a PR!) in the typeshed project. Pyright regularly picks up updates to typeshed stubs.

In the meantime, you can work around the problem by adding an explicit argument for the `result` parameter.

**Comment by wch** [comment]
Thanks! I've filed an issue on typeshed.
